src/alpaca_typer.erl

```diff
@@ -1404,6 +1413,67 @@ type_module_tests([], _, _, Funs) ->
 type_module_tests([#alpaca_test{expression=E}|Rem], Env, _, Funs) ->
     type_module_tests(Rem, Env, typ_of(Env, 0, E), Funs).
 
+%% Here we make a quick pass over each type defined in a module to ensure
+%% types that are members of a type are defined in this module or imported.
+validate_types(#alpaca_module{name=Mod, types=Types, type_imports=Imports}, Mods) ->
+    TypeNames =
+        [N || #alpaca_type{name={_, _, N}} <- Types] ++
+        [N || #alpaca_type_import{type=N} <- Imports],
+    validate_types(Mod, TypeNames, Mods, Types).
+
+validate_types(_ModName, _TypeNames, _Modules, []) ->
+    [];
+validate_types(
+  ModName,
+  TypeNames,
+  Modules,
+  [#alpaca_type{module=MN}=H|T]) when MN =:= undefined; MN =:= ModName ->
+    #alpaca_type{name={_, L, N}, vars=Vs, members=Ms}=H,
+    case [TN || TN <- TypeNames, TN =:= N] of
+        [] ->
+            throw({invalid_type, ModName, L, N});
+        _ ->
+            validate_types(ModName, TypeNames, Modules, Ms),
+            validate_types(ModName, TypeNames, Modules, Vs),
+            validate_types(ModName, TypeNames, Modules, T)
+    end;
+validate_types(MN, Ts, Mods, [#alpaca_type{}=T|Rem]) ->
+    #alpaca_type{name={_, L, N}, module=TargetMod, vars=Vs} = T,
+    case [M || #alpaca_module{name=X}=M <- Mods, X =:= TargetMod] of
+        [] -> 
+            throw({bad_module, MN, L, TargetMod});
+        [#alpaca_module{types=Xs}] ->
+            case [X || #alpaca_type{name={_, _, Y}}=X <- Xs, Y =:= N] of
+                [] -> 
+                    throw({invalid_type, MN, L, N});
+                [_] ->
+                    [] = validate_types(MN, Ts, Mods, Vs),
+                    validate_types(MN, Ts, Mods, Rem)
+            end
+    end;
+ validate_types(MN, TNs, Mods, [{{type_var, _, _}, Typ}|T]) ->
+     validate_types(MN, TNs, Mods, [Typ]),
+     validate_types(MN, TNs, Mods, T);
+ validate_types(MN, TNs, Mods, [#alpaca_constructor{arg=A}|T]) ->
+    validate_types(MN, TNs, Mods, [A]),
+    validate_types(MN, TNs, Mods, T);
+ validate_types(ModName, TypeNames, Mods, [#alpaca_type_tuple{members=Ms}|T]) ->
+    validate_types(ModName, TypeNames, Mods, Ms),
+    validate_types(ModName, TypeNames, Mods, T);
+validate_types(MN, Ts, Mods, [#t_record{members=Ms}|T]) ->
+    MemberTypes = [Type || #t_record_member{type=Type} <- Ms],
+    validate_types(MN, Ts, Mods, MemberTypes ++ T);
+validate_types(M, TNs, Mods, [{t_list, LT}|T]) ->
+    [] = validate_types(M, TNs, Mods, [LT|T]),
+    validate_types(M, TNs, Mods, T);
+validate_types(M, TNs, Mods, [{alpaca_map, KT, VT}|T]) ->
```

Comment:
Here is another example of a duplicate type representation: `{alpaca_map, K, V}` vs `{t_map, K, V}`.